pkg/emitter/nats_emitter_test.go

```diff
@@ -0,0 +1,197 @@
+				"digest": { "sha1": "d6525c840a62b398424a78d792f457477135d0cf" }
+			  }, {
+				"uri": "github_hosted_vm:ubuntu-18.04:20210123.1",
+				"digest": { "sha1": "d6525c840a62b398424a78d792f457477135d0cf" }
+			  }
+			]
+		}
+	}`
+
+	ite6SLSADoc = processor.Document{
+		Blob:   []byte(ite6SLSA),
+		Type:   processor.DocumentITE6SLSA,
+		Format: processor.FormatJSON,
+		SourceInformation: processor.SourceInformation{
+			Collector: "TestCollector",
+			Source:    "TestSource",
+		},
+	}
+)
+
+func TestNatsEmitter_PublishOnEmit(t *testing.T) {
+	expectedDocTree := dochelper.DocNode(&ite6SLSADoc)
+
+	natsTest := nats_test.NewNatsTestServer()
+	url, err := natsTest.EnableJetStreamForTest()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer natsTest.Shutdown()
+
+	ctx := context.Background()
+	jetStream := NewJetStream(url, "", "")
+	ctx, err = jetStream.JetStreamInit(ctx)
+	if err != nil {
+		t.Fatalf("unexpected error initializing jetstream: %v", err)
+	}
+	err = jetStream.RecreateStream(ctx)
+	if err != nil {
+		t.Fatalf("unexpected error recreating jetstream: %v", err)
+	}
+	defer jetStream.Close()
+	err = testPublish(ctx, &ite6SLSADoc)
+	if err != nil {
+		t.Fatalf("unexpected error on emit: %v", err)
+	}
+
+	var cancel context.CancelFunc
+
+	ctx, cancel = context.WithTimeout(ctx, time.Second)
+	defer cancel()
+
+	docChan := make(chan processor.DocumentTree, 1)
+	errChan := make(chan error, 1)
+	defer close(errChan)
+	go func() {
+		errChan <- testSubscribe(ctx, docChan)
+	}()
+
+	numSubscribers := 1
+	subscribersDone := 0
+
+	for subscribersDone < numSubscribers {
+		select {
+		case d := <-docChan:
+			if !dochelper.DocTreeEqual(d, expectedDocTree) {
+				t.Errorf("doc tree did not match up, got\n%s, \nexpected\n%s", dochelper.StringTree(d), dochelper.StringTree(expectedDocTree))
+			}
+		case err := <-errChan:
+			if err != nil && !errors.Is(err, context.DeadlineExceeded) {
+				t.Errorf("nats emitter Subscribe test errored = %v", err)
+			}
+			subscribersDone += 1
+		}
+	}
+
+}
+
+func testPublish(ctx context.Context, d *processor.Document) error {
+	logger := logging.FromContext(ctx)
+	js := FromContext(ctx)
+	docByte, err := json.Marshal(d)
+	if err != nil {
+		return fmt.Errorf("failed marshal of document: %w", err)
+	}
+	_, err = js.Publish(SubjectNameDocCollected, docByte)
+	if err != nil {
+		return fmt.Errorf("failed to publish document on stream: %w", err)
+	}
+	logger.Infof("doc published: %+v", d)
+	return nil
+}
+
+func testSubscribe(ctx context.Context, docChannel chan<- processor.DocumentTree) error {
+	logger := logging.FromContext(ctx)
+	js := FromContext(ctx)
+	id := uuid.NewV4().String()
+	sub, err := js.PullSubscribe(SubjectNameDocCollected, "processor")
+	if err != nil {
+		logger.Errorf("processor subscribe failed: %s", err)
+		return err
+	}
+	for {
+		// if the context is canceled we want to break out of the loop
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		msgs, err := sub.Fetch(1)
+		if err != nil {
```

Comment:
hmm not that it matters too much, but probably useful in the future, is there an error we can check for?

Reply:
added the following to check for the error:
```
		msgs, err := sub.Fetch(1)
		if err != nil {
			if errors.Is(err, nats.ErrTimeout) {
				logger.Infof("[processor: %s] error consuming, backing off for a second: %v", id, err)
				time.Sleep(1 * time.Second)
				continue
			} else {
				return err
			}
		}
```